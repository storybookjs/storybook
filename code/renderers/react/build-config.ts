import type { BuildEntries } from '../../../scripts/build/utils/entry-utils';

const config: BuildEntries = {
  entries: {
    browser: [
      {
        exportEntries: ['.'],
        entryPoint: './src/index.ts',
      },
      {
        exportEntries: ['./preview'],
        entryPoint: './src/preview.tsx',
      },
      {
        exportEntries: ['./entry-preview'],
        entryPoint: './src/entry-preview.tsx',
        dts: false,
      },
      {
        exportEntries: ['./entry-preview-argtypes'],
        entryPoint: './src/entry-preview-argtypes.ts',
        dts: false,
      },
      {
        exportEntries: ['./entry-preview-docs'],
        entryPoint: './src/entry-preview-docs.ts',
        dts: false,
      },
      {
        exportEntries: ['./entry-preview-rsc'],
        entryPoint: './src/entry-preview-rsc.tsx',
        dts: false,
      },
      {
        exportEntries: ['./experimental-playwright'],
        entryPoint: './src/playwright.ts',
      },
    ],
    node: [
      {
        exportEntries: ['./preset'],
        entryPoint: './src/preset.ts',
        dts: false,
      },
    ],
  },
};

export default config;
