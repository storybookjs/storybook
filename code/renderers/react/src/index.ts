/// <reference types="webpack-env" />
import './globals';

export * from './public-types';

export * from './portable-stories';

export * from './preview';

export type { ReactTypes } from './types';

// optimization: stop HMR propagation in webpack

// optimization: stop HMR propagation in webpack
if (typeof module !== 'undefined') {
  module?.hot?.decline();
}
