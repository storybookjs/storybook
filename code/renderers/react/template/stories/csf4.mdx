import * as StoriesModule from './csf4.stories'
import { Meta, Stories } from '@storybook/addon-docs/blocks'

<Meta of={StoriesModule} />

# CSF4 in MDX

<Stories />