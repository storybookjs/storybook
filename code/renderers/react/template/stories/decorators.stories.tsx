import type { FC } from 'react';
import React, { createContext, useContext, useState } from 'react';

import { useParameter } from 'storybook/internal/preview-api';

import type { Meta, StoryObj } from '@storybook/react';

const Component: FC = () => <p>Story</p>;

export default {
  component: Component,
  tags: ['autodocs'],
  decorators: [
    (Story) => (
      <>
        <p>Component Decorator</p>
        <Story />
      </>
    ),
  ],
} as Meta<typeof Component>;

export const All: StoryObj<typeof Component> = {
  decorators: [
    (Story) => (
      <>
        <p>Local Decorator</p>
        <Story />
      </>
    ),
  ],
};

// This story should not error
// See https://github.com/storybookjs/storybook/issues/21900
const TestContext = createContext<boolean>(false);
export const Context: StoryObj<typeof Component> = {
  parameters: { docs: { source: { excludeDecorators: true } } },
  decorators: [
    (Story) => (
      <TestContext.Provider value>
        <Story />
      </TestContext.Provider>
    ),
  ],
  render: function Render(args, context) {
    const value = useContext(TestContext);

    if (!value) {
      throw new Error('TestContext not set, decorator did not run!');
    }
    return <p>Story</p>;
  },
};

/**
 * This story demonstrates is a regression test for this issue with React hooks in Storybook
 * (https://github.com/storybookjs/storybook/issues/29189)
 *
 * Which happened when a decorator was using storybook hooks, and the render react hooks.
 */
export const AllowUseStateInRender: StoryObj = {
  render: () => {
    const [count, setCount] = useState(0);
    const Button = (globalThis as any).Components.Button;
    return <Button onClick={() => setCount(count + 1)} label={`Clicked ${count} times`} />;
  },
  decorators: [
    (storyFn) => {
      useParameter('docs', {});
      return storyFn();
    },
  ],
  play: async ({ canvas, userEvent }) => {
    const button = await canvas.findByText('Clicked 0 times');
    await userEvent.click(button);
    await canvas.findByText('Clicked 1 times');
  },
  tags: ['!vitest'],
};
