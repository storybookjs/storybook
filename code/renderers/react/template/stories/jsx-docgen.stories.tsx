import React, { type FC } from 'react';

const Component: FC<{ children: React.ReactNode }> = ({ children }) => {
  return <div>{children}</div>;
};

export default {
  component: Component,
  tags: ['autodocs'],
};

// This story is used to test whether JSX elements render correctly in autodocs and controls panel
export const Default = {
  args: {
    children: <div>Test</div>,
  },
};
