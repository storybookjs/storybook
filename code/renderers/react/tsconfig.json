{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@storybook/react": ["./src"]
    }
  },
  "include": ["src/**/*", "template/**/*"]
}
