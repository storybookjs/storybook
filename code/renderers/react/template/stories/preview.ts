import { __definePreview } from '@storybook/react';

// This file is used for CSF4 stories
// @ts-expect-error TODO: Fix later with Kasper
export default __definePreview({});
