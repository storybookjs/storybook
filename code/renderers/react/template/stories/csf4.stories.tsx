// @ts-expect-error this will be part of the package.json of the sandbox
import preview from '#.storybook/preview';

const meta = preview.meta({
  component: globalThis.Components.Button,
  args: {
    label: 'Hello world!',
  },
});

export const Story = meta.story({});
