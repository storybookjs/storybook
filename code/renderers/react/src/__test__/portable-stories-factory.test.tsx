// @vitest-environment happy-dom
import { cleanup, render, screen } from '@testing-library/react';
import { afterEach, describe, expect, it, vi } from 'vitest';

import React from 'react';

import {
  CSF2StoryWithLocale,
  CSF3Button,
  CSF3ButtonWithRender,
  CSF3InputFieldFilled,
  CSF3Primary,
  MountInPlayFunction,
  MountInPlayFunctionThrow,
} from './Button.csf4.stories';
import { CSF2Secondary, HooksStory } from './Button.csf4.stories';

afterEach(() => {
  cleanup();
});

describe('renders', () => {
  it('renders primary button', () => {
    render(<CSF2Secondary.Component primary={true}>Hello world</CSF2Secondary.Component>);
    const buttonElement = screen.getByText(/Hello world/i);
    expect(buttonElement).not.toBeNull();
  });

  it('reuses args from composed story', () => {
    render(<CSF2Secondary.Component />);
    const buttonElement = screen.getByRole('button');
    expect(buttonElement.textContent).toEqual(CSF2Secondary.input.args.children);
  });

  it('onclick handler is called', async () => {
    const onClickSpy = vi.fn();
    render(<CSF2Secondary.Component onClick={onClickSpy} />);
    const buttonElement = screen.getByRole('button');
    buttonElement.click();
    expect(onClickSpy).toHaveBeenCalled();
  });

  it('reuses args from composeStories', () => {
    const { getByText } = render(<CSF3Primary.Component />);
    const buttonElement = getByText(/foo/i);
    expect(buttonElement).not.toBeNull();
  });

  it('should render component mounted in play function', async () => {
    await MountInPlayFunction.run();

    expect(screen.getByTestId('spy-data').textContent).toEqual('mockFn return value');
    expect(screen.getByTestId('loaded-data').textContent).toEqual('loaded data');
  });

  it('should throw an error in play function', async () => {
    await expect(() => MountInPlayFunctionThrow.run()).rejects.toThrowError('Error thrown in play');
  });
});

describe('CSF3', () => {
  it('renders with inferred globalRender', () => {
    render(<CSF3Button.Component>Hello world</CSF3Button.Component>);
    const buttonElement = screen.getByText(/Hello world/i);
    expect(buttonElement).not.toBeNull();
  });

  it('renders with custom render function', () => {
    render(<CSF3ButtonWithRender.Component />);
    expect(screen.getByTestId('custom-render')).not.toBeNull();
  });

  it('renders with play function without canvas element', async () => {
    await CSF3InputFieldFilled.run();

    const input = screen.getByTestId('input') as HTMLInputElement;
    expect(input.value).toEqual('Hello world!');
  });

  it('renders with play function with canvas element', async () => {
    let divElement;
    try {
      divElement = document.createElement('div');
      document.body.appendChild(divElement);

      await CSF3InputFieldFilled.run({ canvasElement: divElement });

      const input = screen.getByTestId('input') as HTMLInputElement;
      expect(input.value).toEqual('Hello world!');
    } finally {
      if (divElement) {
        document.body.removeChild(divElement);
      }
    }
  });

  it('renders with hooks', async () => {
    await HooksStory.run();

    const input = screen.getByTestId('input') as HTMLInputElement;
    expect(input.value).toEqual('Hello world!');
  });
});

const testCases = Object.entries({
  CSF2StoryWithLocale,
  CSF3Button,
  CSF3ButtonWithRender,
  CSF3InputFieldFilled,
  CSF3Primary,
  MountInPlayFunction,
}).map(([name, Story]) => [name, Story] as [string, typeof Story]);
it.each(testCases)('Renders %s story', async (_storyName, Story) => {
  await Story.run();
  expect(document.body).toMatchSnapshot();
});
