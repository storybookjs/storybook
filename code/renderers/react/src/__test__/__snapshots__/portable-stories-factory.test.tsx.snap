// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders CSF2StoryWithLocale story 1`] = `
<body>
  <div>
    <p>
      locale: 
    </p>
    <button
      class="storybook-button storybook-button--medium storybook-button--secondary"
      type="button"
    />
  </div>
</body>
`;

exports[`Renders CSF3Button story 1`] = `
<body>
  <div>
    <button
      class="storybook-button storybook-button--medium storybook-button--secondary"
      type="button"
    >
      foo
    </button>
  </div>
</body>
`;

exports[`Renders CSF3ButtonWithRender story 1`] = `
<body>
  <div>
    <div>
      <p
        data-testid="custom-render"
      >
        I am a custom render function
      </p>
      <button
        class="storybook-button storybook-button--medium storybook-button--secondary"
        type="button"
      >
        foo
      </button>
    </div>
  </div>
</body>
`;

exports[`Renders CSF3InputFieldFilled story 1`] = `
<body>
  <div>
    <input
      data-testid="input"
    />
  </div>
</body>
`;

exports[`Renders CSF3Primary story 1`] = `
<body>
  <div>
    <button
      class="storybook-button storybook-button--large storybook-button--primary"
      type="button"
    >
      foo
    </button>
  </div>
</body>
`;

exports[`Renders MountInPlayFunction story 1`] = `
<body>
  <div>
    <div>
      <div
        data-testid="loaded-data"
      >
        loaded data
      </div>
      <div
        data-testid="spy-data"
      >
        mockFn return value
      </div>
    </div>
  </div>
</body>
`;
